### PR TITLE
job-manager: send job annotations to journal instead of publishing

### PR DESCRIPTION
**Pull request by chu11**
per issue #3233, although there were prior discussions in #3208 too.

Note that there are no new "job-info" annotation tests.  All of the annotation tests from before should be sufficient.

### CODE REVIEW

**Comment by garlick** [comment]
Rather than calling `event_batch_annotations()` at every point where the annotations can change, could we just add a flag to `event_job_post_pack()` to indicate that this event is not posted to the permanent job record (but still goes to the event stream?)

I would suggest adding an  `int flags` arg and defining a single `EVENT_POST_INFO_ONLY` flag (or better name).  That way we can add more flags later if needed.

**Comment by chu11** [comment]
> Rather than calling event_batch_annotations() at every point where the annotations can change, could we just add a flag to event_job_post_pack() to indicate that this event is not posted to the permanent job record (but still goes to the event stream?) 

It would work, but I'm not sure if its a net win code wise.

```
    /* O? support in jansson 2.8 */
    if (job->annotations) {
        if (!(entry = eventlog_entry_pack (timestamp,
                                           "annotations",
                                           "{s:o}",
                                           "annotations", job->annotations)))
            goto error;
    }
    else {
        if (!(entry = eventlog_entry_pack (timestamp,
                                           "annotations",
                                           "{s:n}",
                                           "annotations")))
            goto error;
    }
```

the lack of the `O?` (we currently support >= 2.6) support means we'd have to duplicate this branch check in several places.  In addition, once journaling support is added, we'll have to `json_deep_copy()` the `job->annotations` object too.

could give it a shot and see where we end up?

**Comment by garlick** [comment]
It seems like where we should head is every posted event by default goes to both journal and job eventlog, and should be posted with one function call.  The flags addition would distinguish "special" events that only go to the journal.

For the O?, would this work?
```c
event_post ("s:o", o ? o : json_null());
```

I haven't looked too hard but it seems like we are updating the annotations in a half dozen places, but there really should only be like two alloc responses that can update annotations (the annotate one and the success one?)

**Comment by chu11** [comment]
> It seems like where we should head is every posted event by default goes to both journal and job eventlog, and should be posted with one function call.

Ok, I see your point / goal.  Let me see how it looks.

> `event_post ("s:o", o ? o : json_null());`

Yeah, I think that'll work.  I'm not sure why I didn't think of that before.  Hehe, a pattern emerges in code and you just keep on copying it :P

**Comment by chu11** [comment]
re-pushed, with updates given comments above

**Comment by chu11** [comment]
rebased & re-pushed, tweaking for fact I can take advantage of newer jansson version

**Comment by chu11** [comment]
rebased on top of PR #3246, the "misc cleanups" one.

**Comment by chu11** [comment]
re-pushed, re-basing on top of #3246 updates and fixing up some commit messages.

File: src/modules/job-manager/event.c [hunk 1]
```diff
@@ -672,12 +673,12 @@ int event_job_post_pack (struct event *event,
...
     if (!(entry = eventlog_entry_vpack (timestamp, name, context_fmt, ap)))
         return -1;
+    if (event_batch_process_event_entry (event, job, name, entry) < 0)
+        goto error;
     if (event_job_update (job, entry) < 0) // modifies job->state
         goto error;
     if (event_batch_commit_event (event, job, entry) < 0)
         goto error;
-    if (event_batch_process_event_entry (event, job, name, entry) < 0)
-        goto error;
```
Comment (garlick) [inline src/modules/job-manager/event.c]: This reordering probably doesn't belong in this commit?  It should probably have a commit message of its own since the change has a non-obvious (to me) rationale.

File: src/modules/job-manager/event.c [hunk 2]
```diff
@@ -675,6 +675,8 @@ int event_job_post_pack (struct event *event,
         return -1;
     if (event_batch_process_event_entry (event, job, name, entry) < 0)
         goto error;
+    if (EVENT_INFO_ONLY & flags)
+        goto out;
```
Comment (garlick) [inline src/modules/job-manager/event.c:644]: Suggestion: could we call it "journal" instead of "event stream" in the commit message, and maybe call the flag EVENT_JOURNAL_ONLY?   Consistent naming can make things clearer.

**Comment by chu11** [comment]
travis was stuck for some reason, so i rebased to give it a kick

File: src/modules/job-manager/alloc.c [hunk 3]
```diff
@@ -85,6 +85,12 @@ static void interface_teardown (struct alloc *alloc, char *s, int errnum)
                         flux_log_error (ctx->h,
```
Comment (garlick) [inline src/modules/job-manager/alloc.c]: Comment on the commiit message:
```
job-manager: send annotations via streaming RPC

In addition to sending out annotation updates via the job-annotations
event, send annotation updates via the job-manager.events RPC stream.
Send out the events by calling event_job_post_pack() and using the
new EVENT_INFO_ONLY flag.
```

how about

```
job-manager: post job annotations as a journal-only event

In preparation for dropping publication of the job-annotations message,
post job annotations as a journal-only event.  Including annotations in
the journal ensures proper ordering relative to other posted job events.
```

File: src/modules/job-manager/alloc.c [hunk 4]
```diff
@@ -229,6 +235,15 @@ static void alloc_response_cb (flux_t *h, flux_msg_handler_t *mh,
...
+            if (event_job_post_pack (ctx->event,
+                                     job,
+                                     "annotations",
+                                     EVENT_INFO_ONLY,
+                                     "{s:O?}",
+                                     "annotations", job->annotations) < 0)
+                flux_log_error (ctx->h,
+                                "%s: event_job_post_pack: id=%ju",
+                                __FUNCTION__, (uintmax_t)id);
         }
```
Comment (garlick) [inline src/modules/job-manager/alloc.c]: If we're going to log every `event_job_post_pack()` failure in the same way, would it make sense just to log them inside the function?  (Just an idle thought  maybe for future cleanup)

File: src/modules/job-info/job_state.c [hunk 5]
```diff
@@ -35,8 +35,15 @@ struct state_transition {
     double timestamp;
 };
```
Comment (garlick) [inline src/modules/job-info/job_state.c:37]: Comment on the commit message.  Could we just say "job-info: get annotations from the job manager journal"?

File: src/modules/job-manager/alloc.c [hunk 6]
```diff
@@ -81,10 +81,6 @@ static void interface_teardown (struct alloc *alloc, char *s, int errnum)
                 job->alloc_queued = 1;
```
Comment (garlick) [inline src/modules/job-manager/alloc.c:81]: Another on the commit message, how about "job-manager: stop publishing job-annotations message"

The use of "event" gets mighty confusing.

**Review by garlick — APPROVED** [review]
Right!  Sorry that took so long.  A few comments below, mostly superficial and about consistent use of terminology.  Fix up as you see fit and set MWP!   Thanks.

**Comment by garlick** [comment]
Also, could we change the PR title (default release notes entry) to: "job-manager: send job annotations to journal instead of publishing" or similar?  (In the name of consistent terminology and less "event" befuddlement)

File: src/modules/job-manager/alloc.c (hunk 4 above)
Comment (chu11) [reply src/modules/job-manager/alloc.c]: hmmm, the primary reason I do all of that logging is b/c any failures related to annotations shouldn't lead to a teardown of the job-manager, while other `event_job_post_pack()` errors should fail out with a teardown.

I guess that difference wasn't as noticeable before when we used `event_batch_pub_annotations()`.  Yeah, perhaps cleanup not necessary here, but to keep in mind for later.

**Comment by chu11** [comment]
@garlick, thanks!  re-pushed with all the tweaks mentioned above.  I'll leave it for a quick skim if you'd like.  I'll set MWP in the morning.

Before reviewing #3261 / #3254, lemme tweak those commit messages, as we're clearly using the word "journal" more than "events" now.

**Comment by garlick** [comment]
Thanks, looks good!
